vpr/src/pack/cluster.cpp

```diff
@@ -2532,28 +2727,83 @@ static void add_cluster_molecule_candidates_by_attraction_group(t_pb* cur_pb,
                                                                 t_cluster_placement_stats* cluster_placement_stats_ptr,
                                                                 const std::multimap<AtomBlockId, t_pack_molecule*>& atom_molecules,
                                                                 AttractionInfo& attraction_groups,
-                                                                const int feasible_block_array_size) {
+                                                                const int feasible_block_array_size,
+                                                                ClusterBlockId clb_index,
+                                                                std::map<const t_model*, std::vector<t_logical_block_type_ptr>>& primitive_candidate_block_types) {
     auto& atom_ctx = g_vpr_ctx.atom();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    auto cluster_type = cluster_ctx.clb_nlist.block_type(clb_index);
 
     AttractGroupId grp_id = cur_pb->pb_stats->attraction_grp_id;
-    if (grp_id != AttractGroupId::INVALID()) {
-        AttractionGroup group = attraction_groups.get_attraction_group_info(grp_id);
+    if (grp_id == AttractGroupId::INVALID()) {
+        return;
+    }
 
-        for (AtomBlockId blk_id : group.group_atoms) {
-            if (atom_ctx.lookup.atom_clb(blk_id) == ClusterBlockId::INVALID()) {
-                auto rng = atom_molecules.equal_range(blk_id);
+    AttractionGroup& group = attraction_groups.get_attraction_group_info(grp_id);
+    std::vector<AtomBlockId> available_att_group_atoms;
+
+    for (AtomBlockId atom : group.group_atoms) {
+        const auto& atom_model = atom_ctx.nlist.block_model(atom);
+
+        auto itr = primitive_candidate_block_types.find(atom_model);
+        VTR_ASSERT(itr != primitive_candidate_block_types.end());
+        std::vector<t_logical_block_type_ptr>& candidate_types = itr->second;
+
+        //Find whether the logical block type of the current cluster is included in the candidate types of the atom being examined
```

Comment:
also comment that you look at whether the atom is available for this cluster (it's not already packed)